### DISCUSSION TITLE
Managing site-specific configuration in salt

### DISCUSSION THREAD

**Discussion by abesinger** [Security-Onion-Solutions/securityonion#7573] [Unsupported Versions]
I asked this question in #6943, but didn't get a useful answer, I'd going to try rephrasing here.

I've got a distributed install, and I'm going to be adding on the order of 50 nodes to it.  I have some site specific configurations that I want added to all of the nodes, such as shell accounts for admins, some additional packages, etc. I don't want to have to add those by hand -- that's he kind of things that Salt is made for.

My question is, where should I add that into Salt? I don't see any place in the SO Salt config that allows for that. Am I missing something?  I don't want to edit files in `/opt/so/saltstack/default`, because those will get overwritten by soup.  I could copy one of the files into `saltstack/local`, but then it wouldn't get updated by salt and I'd have to keep an eye on the changes.

if SO's salt config doesn't support what I'm asking, what would be a good approach to modifying the Salt config to allow for local changes -- something that I could submit as a PR for a future release? I'm not a Salt expert (I've used other tools such as Ansible and Puppet before), but looking at `saltstack/default/salt/common/init.sls`, I think what could work is addition such as:

{% do allowed_states.append('local_site') %}

and have a dummy (empty) `local_site` file in `saltstack/default/salt` to prevent any `file not found` errors, and then one can create `saltstack/local/salt/local_site` with whatever they need.

Is that the right approach, or is there another way that's better?

**Comment by m0duspwnens** [comment]
@abesinger To accomplish this, you would need to modify the `salt-master` config located on the manager at `/etc/salt/master` and add new [file_roots](https://docs.saltproject.io/en/latest/ref/configuration/master.html#file-roots) and [pillar_roots](https://docs.saltproject.io/en/latest/ref/configuration/master.html#pillar-roots) then within the new `file_roots` directory, you would need to create a top file for that environment and assign the states to the nodes you want via [targeting](https://docs.saltproject.io/en/latest/topics/targeting/index.html).

So for example:
 Create the new directories: `/opt/custom/salt` and `/opt/custom/pillar`.
 Within `/opt/custom/salt`, create your new states.
 Create a new `top.sls` in  `/opt/custom/salt` and assign the states.
```
custom:
  '*_sensor':
    - test
```
 Do the save as above, but for the pillars if needed.
 Add the new directories to the `file_roots` and `pillar_roots` in `/etc/salt/master` and restart the `salt-master` service.
```
file_roots:
  base:
    - /opt/so/saltstack/local/salt
    - /opt/so/saltstack/default/salt
  custom:
    - /opt/custom/salt

pillar_roots:
  base:
    - /opt/so/saltstack/local/pillar
    - /opt/so/saltstack/default/pillar
  custom:
    - /opt/custom/pillar
```
 Go to your minion and run `salt-call state.show_top` to ensure the new states will be applied during a highstate. 
 Run a `salt-call state.highstate -l info` to ensure everything worked as expect.

**Comment by abesinger** [comment]
@m0duspwnens Thanks for the thoughtful answer.

So if i understand correctly, with the approach you suggest, It would help, but I would have to create a pillar file every time I add a node, correct?  The SO installer creates the /opt/so/.../pillar automagically during the install, but it wouldn't create this one?

**Reply by m0duspwnens** [reply]
Not necessarily as that would depend on how the new pillars were assigned to the minions in the new pillar top file.

**Comment by abesinger** [comment]
Thanks, I'll give it a try.
